pkg/app/server/unregisteredappstore/cache.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2022 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unregisteredappstore
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/pipe-cd/pipecd/pkg/cache"
+	"github.com/pipe-cd/pipecd/pkg/cache/rediscache"
+	"github.com/pipe-cd/pipecd/pkg/model"
+	"github.com/pipe-cd/pipecd/pkg/redis"
+)
+
+type Store interface {
+	ListUnregisteredApplications(ctx context.Context, projectID string) ([]*model.ApplicationInfo, error)
+	PutUnregisteredApplications(projectID, pipedID string, apps []*model.ApplicationInfo) error
+}
+
+type store struct {
+	backend redis.Redis
+	logger  *zap.Logger
+}
+
+func NewStore(r redis.Redis, logger *zap.Logger) Store {
+	return &store{
+		backend: r,
+		logger:  logger,
+	}
+}
+
+func (c *store) ListUnregisteredApplications(ctx context.Context, projectID string) ([]*model.ApplicationInfo, error) {
+	key := makeUnregisteredAppsCacheKey(projectID)
+	hc := rediscache.NewHashCache(c.backend, key)
+
+	// pipedToApps assumes to be a map["piped-id"][]byte(slice of *model.ApplicationInfo encoded by encoding/gob)
+	pipedToApps, err := hc.GetAll()
+	if errors.Is(err, cache.ErrNotFound) {
+		return []*model.ApplicationInfo{}, nil
+	}
+
+	if err != nil {
+		c.logger.Error("failed to get unregistered apps", zap.Error(err))
+		return nil, status.Error(codes.Internal, "Failed to get unregistered apps")
```

Comment:
gRPC error is a special error type used for gRPC communication so it should only be returned by gRPC services such as `web-api`, `piped-api`. All other internal packages return only the normal error then the callers (`web-api`, ...) have to convert them to gRPC error.